helix-gateway/src/main/java/org/apache/helix/gateway/channel/HelixGatewayServicePollModeChannel.java

```diff
@@ -20,38 +20,195 @@
  */
 
 import java.io.IOException;
+import org.apache.commons.lang3.NotImplementedException;
 import org.apache.helix.gateway.api.service.HelixGatewayServiceChannel;
 import proto.org.apache.helix.gateway.HelixGatewayServiceOuterClass;
 
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
 
-// TODO: implement this class
+import org.apache.helix.gateway.service.GatewayServiceManager;
+import org.apache.helix.gateway.util.StateTransitionMessageTranslateUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.helix.gateway.channel.GatewayServiceChannelConfig.FileBasedConfigType.*;
+import static org.apache.helix.gateway.util.PollChannelUtil.*;
+
+
+/**
+ * Helix Gateway Service Poll mode implementation.
+ * It periodically polls the current state of the participants and the liveness of the participants.
+ */
 public class HelixGatewayServicePollModeChannel implements HelixGatewayServiceChannel {
+  private static final Logger logger = LoggerFactory.getLogger(HelixGatewayServicePollModeChannel.class);
+  final GatewayServiceManager _manager;
+  final GatewayServiceChannelConfig _config;
+
+  // cluster -> file for user to report shards' current states
+  final String _userCurrentStateFilePath;
+  // cluster -> file path to store the shards' target states
+  final String _targetStateFilePath;
+  final GatewayServiceChannelConfig.ChannelType _participantConnectionStatusChannelType;
+  final GatewayServiceChannelConfig.ChannelType _shardStateChannelType;
+
+  // cluster -> host -> liveness result
+  final Map<String, Map<String, Boolean>> _livenessResults;
+  // cluster -> host -> endpoint for query liveness
+  // It is the file pass if _participantConnectionStatusChannelType is FILE, grpc endpoint if it is GRPC_CLIENT
+  final Map<String, Map<String, String>> _livenessCheckEndpointMap;
+
+  ScheduledExecutorService _scheduler;
+
+  public HelixGatewayServicePollModeChannel(GatewayServiceManager manager, GatewayServiceChannelConfig config) {
+    _manager = manager;
+    _config = config;
+    _scheduler = Executors.newSingleThreadScheduledExecutor();
+    _participantConnectionStatusChannelType = _config.getParticipantConnectionChannelType();
+    _shardStateChannelType = _config.getShardStateChannelType();
+    _livenessCheckEndpointMap = _config.getParticipantLivenessEndpointMap();
+    _userCurrentStateFilePath = _config.getPollModeConfig(PARTICIPANT_CURRENT_STATE_PATH);
+    _targetStateFilePath = _config.getPollModeConfig(SHARD_TARGET_STATE_PATH);
+    _livenessResults = new HashMap<>();
+  }
 
-  public HelixGatewayServicePollModeChannel(GatewayServiceChannelConfig config) {
+  void fetchUpdates() {
```

Comment:
Was it thread safe here? Because the liveness value can be updated while another thread is reading.

Reply:
There is no other thread reading liveness. We only read liveness here to generate connect / disconnect event, witch eventually map to creating/deleting Helix participant.

Reply:
Add scope of the function, make protected explicitly.

Also add java doc to describe this. It can prevent future development to be more careful